{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 3 - Machine Learning, Fall 2018"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This project is about using a hidden Markov model for gene finding in prokaryotes as explained in class and explored in the exercises in week 46.\n",
    "\n",
    "We suggest that you do the project in groups of 2-3 students. The project is mandatory and each group must hand in a report describing their work and results as outlined in section Report below. Handin the report in pdf-format via Blackboard no later than **Friday, November 23, 2018, at 16:00**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You are give a data set containing 10\n",
    "[Staphylococcus](http://en.wikipedia.org/wiki/Staphylococcus) genomes,\n",
    "each containing several genes (i.e. substring) obeying the \"gene syntax\" explained in class.\n",
    "The genomes are between 1.8 million and 2.8 million nucleotides.\n",
    "For 5 of the genomes, you are also given the location of the genes.\n",
    "For the remaining 5 genomes, you only know that they contain genes according to the \"gene syntax\". (You have already explored two of the genomes in the practical exercises in week 45.)\n",
    "\n",
    "The genomes and their annontations are given in\n",
    "[FASTA format](http://en.wikipedia.org/wiki/Fasta_format).\n",
    "There are several Python libraries for reading sequences in this format (use Google to find one),\n",
    "but it is also easy to read it directly,\n",
    "and you are welcome to use the simple reader function `read_fasta_file` below.\n",
    "The data is:\n",
    "\n",
    "  * The files [genome1.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome1.fa),\n",
    "    [genome2.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome2.fa),\n",
    "    [genome3.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome3.fa),\n",
    "    [genome4.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome4.fa),\n",
    "    [genome5.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome5.fa),\n",
    "    [genome6.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome6.fa),\n",
    "    [genome7.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome7.fa),\n",
    "    [genome8.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome8.fa),\n",
    "    [genome9.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome9.fa), and\n",
    "    [genome10.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome10.fa) contain the 10 genomes.\n",
    "  * The files [true-ann1.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/true-ann1.fa),\n",
    "    [true-ann2.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/true-ann2.fa),\n",
    "    [true-ann3.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/true-ann3.fa),\n",
    "    [true-ann4.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/true-ann4.fa), and\n",
    "    [true-ann5.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/true-ann5.fa) contain the annotation\n",
    "    of the first 5 genomes. The annotation is given in FASTA format as a sequence\n",
    "    over the symbols `N`, `C`, and `R`. The symbol `N`, `C`, or `R` at position _i_ in\n",
    "    true-ann_k_.fa gives the \"state\" of the nucleotide at position _i_ in\n",
    "    genome_k_.fa. `N` means that the nucleotide is non-coding. `C` means that the\n",
    "    nucleotide is coding and part of a gene in the direction from left to right. `R`\n",
    "    means that the nucleotide is coding and part of gene in the reverse direction\n",
    "    from right to left. (This is exactly as in the pratical exercises in week 45.)   \n",
    "  * All the above files are available in [data-handin3.zip](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/data-handin3.zip)\n",
    "\n",
    "By simple reading of the annotations and the corresponding genomes, it is\n",
    "clear that fx several start-codons are possible (and not only atg as modeled\n",
    "in class).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Reads the given FASTA file f and returns a dictionary of sequences.\n",
    "\n",
    "    Lines starting with ';' in the FASTA file are ignored.\n",
    "    \"\"\"\n",
    "    sequences_lines = {}\n",
    "    current_sequence_lines = None\n",
    "    with open(filename) as fp:\n",
    "        for line in fp:\n",
    "            line = line.strip()\n",
    "            if line.startswith(';') or not line:\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                sequence_name = line.lstrip('>')\n",
    "                current_sequence_lines = []\n",
    "                sequences_lines[sequence_name] = current_sequence_lines\n",
    "            else:\n",
    "                if current_sequence_lines is not None:\n",
    "                    current_sequence_lines.append(line)\n",
    "    sequences = {}\n",
    "    for name, lines in sequences_lines.items():\n",
    "        sequences[name] = ''.join(lines)\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function is used as follows. First, we write an example FASTA file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = '''\n",
    ";Example FASTA file\n",
    ">genomeA\n",
    "CGATTAAAGA\n",
    "TAGAAATACA\n",
    ">annotationA\n",
    "CCCCCCNNNN\n",
    "NNNNRRRRRR\n",
    "'''\n",
    "with open('test.fa', 'w') as fp:\n",
    "    fp.write(s)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we read the file with `read_fasta_file`, we get a dictionary with the keys `genomeA` and `genomeB` corresponding to the contents of the above file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'genomeA': 'CGATTAAAGATAGAAATACA', 'annotationA': 'CCCCCCNNNNNNNNRRRRRR'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = read_fasta_file('test.fa')\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGATTAAAGATAGAAATACA\n"
     ]
    }
   ],
   "source": [
    "print(x['genomeA'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem\n",
    "\n",
    "Your task is to predict the gene structure of the 5 unannotated genomes, i.e.\n",
    "[genome6.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome6.fa),\n",
    "[genome7.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome7.fa),\n",
    "[genome8.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome8.fa),\n",
    "[genome9.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome9.fa), and\n",
    "[genome10.fa](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/genome10.fa).\n",
    "\n",
    "As explained in class, predicting gene structure using a HMM involves:\n",
    "\n",
    "  * Deciding on an initial model structure, i.e. the number of hidden states\n",
    "    and which transitions and emission should have a fixed probability (e.g. 0\n",
    "    for \"not possible\", or 1 for \"always the case\").\n",
    "    You might get inspiration from, or even use, one of the model structures discussed in class.\n",
    "    However, be aware that the models presented in class assume start-codon atg.\n",
    "    This is not the case in the presented data, where multiple start-codons are possible.\n",
    "\n",
    "    To predict the full gene structure, you must decide how to handle to fact that\n",
    "    the genomes have genes in both directions (i.e. a nucleotide can be `C` or `R`).\n",
    "    You can e.g. choose to make a HMM which only models genes in one direction and\n",
    "    then use it twice to predict the genes in each direction (beware that start-\n",
    "    and stop-codons look different in the reversed direction, unless you make the\n",
    "    reverse-complement of the input string), or you can make a model which models\n",
    "    genes in both directions.\n",
    "\n",
    "\n",
    "  * Tune model parameters by training, i.e. set the non-fixed emission and transition probabilities.\n",
    "    This can be done by several methods as explained in class:\n",
    "    (1) \"By counting\" using the 5 genomes with know gene structure because you\n",
    "    know the underlying hidden state for each observation (i.e. symbol) in these\n",
    "    sequences, or\n",
    "    (2) by Viterbi- or EM-training using all the 10 genomes,\n",
    "    including the 5 genomes with unknown gene structure.\n",
    "    You can of course also combine the two approaches.\n",
    "    You can evaluate the performance of your gene finder by computing the\n",
    "    approximate correlation (AC) between your predictions and\n",
    "    the true structure using this small python program\n",
    "    [compare_anns.py](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/compare_anns.py)\n",
    "    (see [this paper](http://www.sciencedirect.com/science/article/pii/S0888754396902980)\n",
    "    for details).\n",
    "\n",
    "    In this project, training by counting is mandatory, while Viterbi- or\n",
    "    EM-training is optional. If you implement EM-training you should of course be\n",
    "    aware of the precision problems of the forward- and backward-algorithms.\n",
    "\n",
    "\n",
    "  * Use your best model to predict the gene structure for the 5 unannotated\n",
    "    genomes using the Viterbi algorithm with subsequent backtracking.\n",
    "    I.e. for each unannotated genome you must find a most likely sequence of\n",
    "    states in your model generating it, and convert this sequence of states into a\n",
    "    FASTA file giving the annotation of each nucleotide as `N`, `C`, or `R`.\n",
    "    You should make files, `pred-ann6.fa`, ..., `pred-ann10.fa`, giving the\n",
    "    annotation for the 5 unannotated genomes. \n",
    "    \n",
    "    You can use the www-service [GeneFinder Verifier](https://services.birc.au.dk/genefinder-verifier/) to compare your predictions on Genome 6-10 against their true structures.\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Report\n",
    "\n",
    "You (i.e. your group) must hand in a report (in pdf-format) describing your\n",
    "work and results and a zip-archive of the files `pred-ann6.fa` and\n",
    "`pred-ann7.fa`,..., `pred-ann10.fa` containing your predictions on the 5\n",
    "genomes with unknown structure. The files _must_ be in FASTA format similar to\n",
    "the annotation of the other genomes. Handin the report and zip-archive via\n",
    "Blackboard no later than **Friday, November 23, 2018, at 16:00**. Your report should\n",
    "be no more than 3 pages and clearly state who is in the group. It must cover:\n",
    "\n",
    "  * The status of the work, i.e. does it work, if not, then why.\n",
    "\n",
    "\n",
    "  * An explanation and illustration of your model structure (i.e. a transition\n",
    "    diagram). This should include an explanation of how you model that there\n",
    "    can be several start and stop codons, and an explanation of how you have\n",
    "    trained your model. For training by counting you should comment on how you have\n",
    "    translated the given annotations of `C`, `N`, and `R`'s into corresponding sequences\n",
    "    of hidden states. The transition-diagram should contain all non-0 probabilities\n",
    "    as computed by your training.\n",
    "\n",
    "\n",
    "  * An explanation of how you have predicted the gene structure for the unannotated sequences.\n",
    "\n",
    "\n",
    "  * An evaluation of the performance of your gene predictor.\n",
    "    You must make a 5-fold cross validation on the 5 genomes with known structure:\n",
    "\n",
    "    For each genome 1 to 5, you train your model by training-by-counting on the\n",
    "    remaining 4 genomes, and predict the gene structure on the genome you picked.\n",
    "    You compute and report the approximate correlation coefficient (AC) between\n",
    "    your prediction and the true annotation using the small python program\n",
    "    [compare_anns.py](https://users-cs.au.dk/cstorm/courses/ML_e18/projects/handin3/compare_anns.py).\n",
    "    Include a table with the computed ACs in your report.\n",
    "    \n",
    "    \n",
    "  * The result of comparing your predictions on the 5 genomes with unknown structure against their true structures via the www-service [GeneFinder Verifer](https://services.birc.au.dk/genefinder-verifier/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import math\n",
    "\n",
    "def read_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Reads the given FASTA file f and returns a dictionary of sequences.\n",
    "\n",
    "    Lines starting with ';' in the FASTA file are ignored.\n",
    "    \"\"\"\n",
    "    sequences_lines = {}\n",
    "    current_sequence_lines = None\n",
    "    with open(filename) as fp:\n",
    "        for line in fp:\n",
    "            line = line.strip()\n",
    "            if line.startswith(';') or not line:\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                sequence_name = line.lstrip('>')\n",
    "                current_sequence_lines = []\n",
    "                sequences_lines[sequence_name] = current_sequence_lines\n",
    "            else:\n",
    "                if current_sequence_lines is not None:\n",
    "                    current_sequence_lines.append(line)\n",
    "    sequences = {}\n",
    "    for name, lines in sequences_lines.items():\n",
    "        sequences[name] = ''.join(lines)\n",
    "    return sequences\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class HMM:\n",
    "    \"\"\"\n",
    "        Make init and transform to np array. \n",
    "    \"\"\"\n",
    "    def __init__(self, init_probs, trans_probs, emission_probs):\n",
    "        self.init_probs = np.array(init_probs)\n",
    "        self.trans_probs = np.array(trans_probs)\n",
    "        self.emission_probs = np.array(emission_probs)\n",
    "\n",
    "    def log(self, x):\n",
    "        if x == 0:\n",
    "            return float('-inf')\n",
    "        return math.log(x)\n",
    "\n",
    "    \"\"\"\n",
    "        Make mapping from: \n",
    "            obs_example = 'GTTTCCCAGTGTATATCGAGGGATACTACGTGCATAGTAACATCGGCCAA'\n",
    "        to:\n",
    "            obs_example_trans= [2,3,3,3,1,1,1,0,2,3,2,3,0,3,0,3,1,2,0,2,2,2,0,3,0,1,3,0,1,2,3,2,1,0,3,0,2,3,0,0,1,0,3,1,2,2,1,1,0,0]\n",
    "    \"\"\"\n",
    "    def translate_observations_to_indices(self, obs):\n",
    "        mapping = {'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
    "\n",
    "        return [mapping[symbol.lower()] for symbol in obs]\n",
    "\n",
    "\n",
    "    def translate_indices_to_observations(self, indices):\n",
    "        mapping = ['a', 'c', 'g', 't']\n",
    "\n",
    "        return ''.join(mapping[idx] for idx in indices)\n",
    "\n",
    "\n",
    "    def translate_path_to_indices(self, path):\n",
    "\n",
    "        return list(map(lambda x: int(x), path))\n",
    "\n",
    "    def translate_indices_to_path(self, indices):\n",
    "        \n",
    "        return ''.join([str(i) for i in indices])\n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "        validate hmm as follows.\n",
    "        1. The initial probabilities must sum to 1.\n",
    "        2. Each row in the matrix of transition probabilities must sum to 1.\n",
    "        3. Each row in the matrix of emission probabilities must sum to 1.\n",
    "        4. All numbers should be between 0 and 1, inclusive.\n",
    "    \"\"\"\n",
    "    \n",
    "    def validate(self):\n",
    "        # 1. \n",
    "        if self.init_probs.sum() != 1:\n",
    "            return False\n",
    "        \n",
    "        # 2.\n",
    "        row_sums2 = self.trans_probs.sum(axis=1)\n",
    "        row_ones2 = np.ones(row_sums2.shape)\n",
    "        if not np.allclose(row_sums2, row_ones2):\n",
    "            return False\n",
    "        \n",
    "        # 3. \n",
    "        row_sums3 = self.emission_probs.sum(axis=1)\n",
    "        row_ones3 = np.ones(row_sums3.shape)\n",
    "        if not np.allclose(row_sums3, row_ones3):\n",
    "            return False\n",
    "\n",
    "        # 4. Are they in range [0:1]\n",
    "        init_range = np.all((self.init_probs >= 0) & (self.init_probs <= 1))\n",
    "        trans_range = np.all((self.trans_probs >= 0) & (self.trans_probs <= 1))\n",
    "        emission_range = np.all((self.emission_probs >= 0) & (self.emission_probs <= 1))\n",
    "        \n",
    "        if not (init_range and trans_range and emission_range):\n",
    "            return False\n",
    "        \n",
    "        # We have passed all tests. It is a valid HMM\n",
    "        return True \n",
    "\n",
    "\n",
    "    \"\"\"\n",
    "    Returns the log transformed joint probability of x and z\n",
    "    \"\"\"\n",
    "    def joint_prob_log(self, x, z):\n",
    "        logp = self.log(self.init_probs[z[0]]) + self.log(self.emission_probs[z[0]][x[0]])\n",
    "        for i in range(1, len(x)):\n",
    "            logp = logp + self.log(self.trans_probs[z[i-1]][z[i]]) + self.log(self.emission_probs[z[i]][x[i]])\n",
    "        return logp\n",
    "\n",
    "\n",
    "    \n",
    "\"\"\"\n",
    "    ########## HELPER FUNCTIONS ###############3\n",
    "\"\"\" \n",
    "\n",
    "def make_table(m, n):\n",
    "    \"\"\"\n",
    "        Make a table with `m` rows and `n` columns filled with zeros.\n",
    "    \"\"\"\n",
    "    return [[0] * n for _ in range(m)]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def main():\n",
    "    # start-codon + inner + stop-codon = 9+9+3 = 21\n",
    "    # Coding and reverese coding = 42\n",
    "    # Init state = 43\n",
    "    K = 43\n",
    "    def fromto(source, target):\n",
    "        return [int(i==target) for i in range(K)]\n",
    "\n",
    "    \n",
    "    def from_distribute_to(source, targetlist):\n",
    "        val = 1.0/len(targetlist)\n",
    "        result = [0 for i in range(K)]\n",
    "        for j in range(len(targetlist)):\n",
    "            result[targetlist[j]] = val\n",
    "        return result\n",
    "\n",
    "    init_probs = fromto(0, 0)  #Only start in state 0. Hacky reuse of func :) \n",
    "\n",
    "    init_trans_probs = [\n",
    "        from_distribute_to(0, [1,4,7,34,37,40]),         # 0 -> 1 | 0 -> 4 | 0 -> 7 | 0 -> 34 | 0 -> 37 | 0 -> 40\n",
    "        #Forward\n",
    "        #start-codon\n",
    "        fromto(1,2), fromto(2,3), fromto(3,10),         # 1 -> 2 -> 3 -> 10\n",
    "        fromto(4,5), fromto(5,6), fromto(6,10),         # 4 -> 5 -> 6 -> 10\n",
    "        fromto(7,8), fromto(8,9), fromto(9,10),         # 7 -> 8 -> 9 -> 10\n",
    "        #inner                    \n",
    "        fromto(10,11),                                  # 10 \\> 11 \n",
    "        fromto(11,12),                                  # 11 \\> 12\n",
    "        from_distribute_to(12, [10, 15, 18, 21]),        # 12 -> 10 | 12 -> 15 | 12 -> 18 | 12 -> 21\n",
    "        #stop-codon\n",
    "        fromto(13, 0), fromto(14, 13), fromto(15, 14),  # 0 <- 13 <- 14 <- 15\n",
    "        fromto(16, 0), fromto(17, 16), fromto(18,17),   # 0 <- 16 <- 17 <- 18\n",
    "        fromto(19, 0), fromto(20, 19), fromto(21,20),   # 0 <- 19 <- 20 <- 21\n",
    "\n",
    "        #Reverse\n",
    "        #stop-codon        \n",
    "        fromto(22, 0), fromto(23, 22), fromto(24,23),   # 0 <- 22 <- 23 <- 24\n",
    "        fromto(25, 0), fromto(26, 25), fromto(27,26),   # 0 <- 25 <- 26 <- 27\n",
    "        fromto(28, 0), fromto(29, 28), fromto(30,29),   # 0 <- 28 <- 29 <- 30\n",
    "        #inner        \n",
    "        from_distribute_to(31, [33,30,27,24]),          # 31 -> 33 | 31 -> 30 | 31 -> 27 | 31 -> 24                                  \n",
    "        fromto(32, 31),                                 # 32 \\> 31\n",
    "        fromto(33, 32),                                 # 33 \\> 32\n",
    "        #start-codon        \n",
    "        fromto(34, 35), fromto(35, 36), fromto(36, 33), # 34 -> 35 -> 36 -> 33\n",
    "        fromto(37, 38), fromto(38, 39), fromto(39, 33), # 37 -> 38 -> 39 -> 33\n",
    "        fromto(40, 41), fromto(41, 42), fromto(42, 33)  # 40 -> 41 -> 42 -> 33\n",
    "    ]\n",
    "    \n",
    "\n",
    "    init_emission_probs = [\n",
    "        #   A     C     G     T\n",
    "        [0.25, 0.25, 0.25, 0.25],   #init\n",
    "\n",
    "        #Forward\n",
    "        [1, 0, 0, 0],   #start1     1\n",
    "        [0, 0, 0, 1],   #start1     2\n",
    "        [0, 0, 1, 0],   #start1     3\n",
    "        [0, 0, 1, 0],   #start2     4\n",
    "        [0, 0, 0, 1],   #start2     5\n",
    "        [0, 0, 1, 0],   #start2     6\n",
    "        [0, 0, 0, 1],   #start3     7\n",
    "        [0, 0, 0, 1],   #start3     8\n",
    "        [0, 0, 1, 0],   #start3     9\n",
    "        [0.25, 0.25, 0.25, 0.25],   #inner\n",
    "        [0.25, 0.25, 0.25, 0.25],   #inner\n",
    "        [0.25, 0.25, 0.25, 0.25],   #inner\n",
    "        [0, 0, 0, 1],   #stop1      13\n",
    "        [1, 0, 0, 0],   #stop1      14\n",
    "        [1, 0, 0, 0],   #stop1      15\n",
    "        [0, 0, 0, 1],   #stop2      16\n",
    "        [1, 0, 0, 0],   #stop2      17\n",
    "        [0, 0, 1, 0],   #stop2      18\n",
    "        [0, 0, 0, 1],   #stop3      19\n",
    "        [0, 0, 1, 0],   #stop3      20\n",
    "        [1, 0, 0, 0],   #stop3      21\n",
    "\n",
    "        # Reverse\n",
    "        # A  C  G  T\n",
    "        [0, 1, 0, 0],   #stop1\n",
    "        [1, 0, 0, 0],   #stop1\n",
    "        [0, 0, 0, 1],   #stop1\n",
    "        [0, 1, 0, 0],   #stop2\n",
    "        [1, 0, 0, 0],   #stop2\n",
    "        [0, 1, 0, 0],   #stop2\n",
    "        [0, 1, 0, 0],   #stop3\n",
    "        [1, 0, 0, 0],   #stop3\n",
    "        [1, 0, 0, 0],   #stop3\n",
    "        [0.25, 0.25, 0.25, 0.25],   #inner\n",
    "        [0.25, 0.25, 0.25, 0.25],   #inner\n",
    "        [0.25, 0.25, 0.25, 0.25],   #inner\n",
    "        [0, 0, 0, 1],   #start1\n",
    "        [0, 0, 0, 1],   #start1\n",
    "        [1, 0, 0, 0],   #start1\n",
    "        [0, 1, 0, 0],   #start2\n",
    "        [0, 0, 0, 1],   #start2\n",
    "        [1, 0, 0, 0],   #start2\n",
    "        [0, 0, 0, 1],   #start3\n",
    "        [0, 1, 0, 0],   #start3\n",
    "        [1, 0, 0, 0],   #start3\n",
    "    ]\n",
    "\n",
    "    #x = read_fasta_file('genome1.fa')\n",
    "    \n",
    "    hmm = HMM(init_probs, init_trans_probs, init_emission_probs)\n",
    "    isvalid = hmm.validate()\n",
    "    \n",
    "    print(isvalid)\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
